### PR TITLE
feat: add the number of merge buffers used for druid emitter

### LINKED ISSUE #4345: Metrics for merge buffer usage
Would be nice to have two metrics for merge buffers, so operators can know if they are frequently maxed out on merge buffers.

- number of merge buffers currently allocated. I imagine it'd emit the instantaneous number of merge buffers used, once a minute. So it might miss brief spikes but should catch prolonged ones.
- amount of time spent waiting for a merge buffer. I imagine this would be per query.

### PR DESCRIPTION
**Pull request by TessaIO**
Fixes #4345 .

#### Release note

<hr>
Add the current number of merge buffers used for historical, broker, and real-time.

<hr>

This PR has:

- [x] been self-reviewed.
   - [x] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: docs/operations/metrics.md [hunk 1]
```diff
@@ -61,6 +61,7 @@ Most metric values reset each emission period, as specified in `druid.monitoring
 |`query/success/count`|Number of queries successfully processed.|This metric is only available if the `QueryCountStatsMonitor` module is included.| |
 |`query/failed/count`|Number of failed queries.|This metric is only available if the `QueryCountStatsMonitor` module is included.| |
 |`query/interrupted/count`|Number of queries interrupted due to cancellation.|This metric is only available if the `QueryCountStatsMonitor` module is included.| |
+|`query/merge/buffersUsed`|number of merge buffers allocated to broker while performing groupBy merge queries.|This metric is only available if the `QueryCountStatsMonitor` module is included.| |
```
Comment (LakshSingla) [inline docs/operations/metrics.md:64]: This shouldn't be "allocated to broker". As per my perception, allocated to the broker means the total number of merge buffers present there. Also, there is no such thing as "groupBy merge query". 
```suggestion
|`query/merge/buffersUsed`|Number of merge buffers used up to merge the results of group by queries.|This metric is only available if the `QueryCountStatsMonitor` module is included.| |
```

File: processing/src/test/java/org/apache/druid/query/TestBufferPool.java [hunk 2]
```diff
@@ -147,4 +147,10 @@ public Collection<RuntimeException> getOutstandingExceptionsCreated()
   {
     return takenFromMap.values();
   }
+
+  @Override
+  public int getUsedBufferCount()
+  {
+    return Integer.MAX_VALUE;
+  }
```
Comment (LakshSingla) [inline processing/src/test/java/org/apache/druid/query/TestBufferPool.java:155]: We can probably have the correct implementation for this class, even if it is a test utility.

**Review by LakshSingla — COMMENTED** [review]
Thanks for the first contribution! 

The code style would be failing on a few files. Checkout 
a) https://github.com/apache/druid/blob/master/dev/intellij-setup.md#code-style to setup the editor to lint the code correctly
b) Running `mvn checkstyle:checkstyle` would help in confirming if the checkstyle would fail on the PR or not.

Regarding the PR:
a. I think there should be a `query/merge/totalBuffers` metric as well. 
b. Apart from this, there can be a wait time associated with each query - the total time required to acquire the merge buffers - this should be a simple change given that now we acquire all the merge buffers in a single place 
c. There should be tests for the changes made in the PR.

**Comment by TessaIO** [comment]
Thanks for the review @LakshSingla.
I addressed all of your comments. PTAL when you have time.

> b. Apart from this, there can be a wait time associated with each query - the total time required to acquire the merge buffers - this should be a simple change given that now we acquire all the merge buffers in a single place

I didn't really get what needed to be done here. Can you elaborate more on this?

**Review by cryptoe — APPROVED** [review]

**Comment by LakshSingla** [comment]
@TessaIO Aplogy for getting back late on this PR. The PR LGTM, can you please address the test failures. 

My original comment was related to adding more metrics around this area, such as the elapsed time between the query requesting the merge buffer, and the query getting allotted the merge buffer. That can be a part of subsequent PRs.

**Comment by TessaIO** [comment]
@LakshSingla @cryptoe any idea on how to improve coverage to make the pipeline pass?

**Comment by cryptoe** [comment]
@TessaIO Could you please rebase the patch. I cannot pull the logs from the older CI runs to see the failures.

**Comment by TessaIO** [comment]
@cryptoe done!

**Comment by cryptoe** [comment]
@TessaIO Could you please fix the CI.

**Pull request closed without merging**